networking.go

```diff
@@ -133,6 +136,48 @@ func startStatsdTCP(s *Server, addr *net.TCPAddr, packetPool *sync.Pool) net.Add
 	return listener.Addr()
 }
 
+// startStatsdUnix starts listening for datagram statsd metric packets
+// on a UNIX domain socket address. It does so until the
+// server's shutdown socket is closed. startStatsdUnix returns a channel
+// that is closed once the listening connection has terminated.
+func startStatsdUnix(s *Server, addr *net.UnixAddr, packetPool *sync.Pool) (<-chan struct{}, net.Addr) {
+	done := make(chan struct{})
+	if addr.Network() != "unixgram" {
+		panic(fmt.Sprintf("Can't listen for statsd metrics on %v: only udp:// and unixgram:// addresses are supported", addr))
+	}
+	// ensure we are the only ones locking this socket:
+	lock := acquireLockForSocket(addr)
+
+	conn, err := net.ListenUnixgram(addr.Network(), addr)
+	if err != nil {
+		panic(fmt.Sprintf("Couldn't listen on UNIX socket %v: %v", addr, err))
+	}
+
+	// Make the socket connectable by everyone with access to the socket pathname:
+	err = os.Chmod(addr.String(), 0666)
+	if err != nil {
+		panic(fmt.Sprintf("Couldn't set permissions on %v: %v", addr, err))
+	}
+	go func() {
```

Comment:
I think this might be one layer of `go func()` too many - the goroutine does not do anything other than spawn more goroutines.

Reply:
Thanks, done